File: HashUtilities.cpp
///////////////////////////////////////////////////////////////////////////////
// Author: Katie Garner
// Description: This is the implementation file for password hashing utilities
//              for the COSC412 Project.
///////////////////////////////////////////////////////////////////////////////
#include "HashUtilities.h"
#include <iomanip>
#include <sstream>
#include <ctime>
#include <cstdlib>

const string HashUtilities::pepper = "Sp1c3y!";

// Description: Generates a random salt, stores it in outSalt, and returns
//              the hashed password string.
string HashUtilities::hashPassword(const string& password, string& outSalt) {
    // Seed the random number generator once
    static bool seeded = false;
    if (!seeded) {
        srand(static_cast<unsigned int>(time(0)));
        seeded = true;
    }

    outSalt = generateSalt(SALT_LENGTH);
    return doHash(outSalt, password);
}

// Description: Hashes a password using a provided salt. Used for login
//              verification when the stored salt is already known.
string HashUtilities::hashWithSalt(const string& password, const string& salt) {
    return doHash(salt, password);
}

// Description: Core DJB2 hash function.
unsigned long HashUtilities::djb2(const string& input) {
    unsigned long hash = 5381;

    for (char c : input) {
        hash = ((hash << 5) + hash) + c;   // hash * 33 + c
    }

    return hash;
}

// Description: Generates a random salt string.
string HashUtilities::generateSalt(int length) {
    const string charSet =
        "abcdefghijklmnopqrstuvwxyz"
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "0123456789!@#$%^&*()";

    string salt = "";

    for (int i = 0; i < length; i++) {
        salt += charSet[rand() % charSet.length()];
    }

    return salt;
}

// Description: Converts an unsigned long hash value to hexadecimal.
string HashUtilities::toHex(unsigned long hash, int width) {
    stringstream hashCode;
    hashCode << hex << setw(width) << setfill('0') << hash;
    return hashCode.str();
}

// Description: Performs the actual salted + peppered hash calculation.
string HashUtilities::doHash(const string& salt, const string& password) {
    return toHex(djb2(salt + password + pepper));
}

File: hash_utilities.py
import random

# Hash utility class for generating salts and hashing passwords
class HashUtilities:
    SALT_LENGTH = 16
    PEPPER = "Sp1c3y!"

    @staticmethod
    def hash_password(password: str):
        """
        Generates a random salt and returns:
        (hashed_password, generated_salt)
        """
        # Create a new random salt for this password
        out_salt = HashUtilities.generate_salt(HashUtilities.SALT_LENGTH)

        # Hash password using the generated salt
        hashed_password = HashUtilities.do_hash(out_salt, password)
        return hashed_password, out_salt

    @staticmethod
    def hash_with_salt(password: str, salt: str) -> str:
        """
        Hashes a password using an already-known salt.
        Used for login verification.
        """
        return HashUtilities.do_hash(salt, password)

    @staticmethod
    def djb2(input_string: str) -> int:
        """
        Core DJB2 hash function.
        """
        hash_value = 5381

        # Update hash value using each character
        for char in input_string:
            hash_value = ((hash_value << 5) + hash_value) + ord(char)

        return hash_value

    @staticmethod
    def generate_salt(length: int = SALT_LENGTH) -> str:
        """
        Generates a random salt string.
        """
        # Allowed characters for the random salt
        char_set = (
            "abcdefghijklmnopqrstuvwxyz"
            "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
            "0123456789!@#$%^&*()"
        )

        salt = ""

        # Build salt one random character at a time
        for _ in range(length):
            salt += random.choice(char_set)

        return salt

    @staticmethod
    def to_hex(hash_value: int, width: int = 8) -> str:
        """
        Converts a hash value to hexadecimal.
        """
        return format(hash_value, "x").zfill(width)

    # Combine salt, password, and pepper, then return the final hash
    @staticmethod
    def do_hash(salt: str, password: str) -> str:
        """
        Performs the actual salted + peppered hash calculation.
        """
        # Combine values before hashing
        combined = salt + password + HashUtilities.PEPPER

        # Hash combined string and convert to hex
        return HashUtilities.to_hex(HashUtilities.djb2(combined))